Full-stack Todo App (Workshop bounty) (#1568)
hackclub/hackclub: workshops/full-stack_todo_app/README.md

CRUD stands for Create, Read, Update, and Delete.

These are the four most basic operations that can be performed with most traditional database systems, and they are the backbone for interacting with any database. 

![CRUD](https://cloud-9uzz77u17.vercel.app/0crud_operations.png)

## Creating Todo App

Let's create the interface of the Todo App to perform CRUD operations.

Create a new folder and name it as `components`. All of our Todo App components will be in this folder.

### Inputting a Todo

Create a new file in the `components` folder and name it `Input.js`. Import `React` into the component and create the functional component, and export it.

```jsx
import React from 'react'

function Input(){
  // code goes here
  }
 
 export default Input
```

Also, import the `Firebase.js` file you created before at the top to use Firebase inside this component.

```jsx
import firebse from '../Firebase'
```

Create a `div` with a class of `input` inside the component.  

```jsx
import React from 'react'
import firebase from '../Firebase'

function Input(){
  return(
    <div className="input">

>       // code goes here

Reviewer: jsx comments are not starting with "//", try to change it or it might confuse the people doing this workshop.

Author: I tried but the markdown preview was not supporting
The first `{` is not  commented.
```jsx
import React from 'react'

function Input(){
  {// code goes here}
  }
 
 export default Input
```

    </div>
  )
}

export default Input
```

Add an `input` and a button `Add Todo` with a class of `add-btn` inside the `div` element.

```html
<div className="input">
  <input type="text" placeholder="Enter a Todo..." />
  <button className="add-btn">
  <i className="fa fa-plus-circle" aria-hidden="true"></i><span id="btn-text">Add Todo</span> 
  </button>
</div>
```

> Here, `<i className="fa fa-plus-circle">` is the [plus icon](https://fontawesome.com/v4.7.0/icon/plus-circle) from [font awesome](https://fontawesome.com/v4.7.0/icons/) for styling. 

Reviewer: can u explain where we imported font-awesome and sme stuff about it? coz if people directly see font-awesome being used without any information, they'll get confused

Create a state value and update it every time the input changes using the `handleChange` function.

```jsx
import React, { useState } from 'react'  // <-------------- importing useState hook
import firebase from '../Firebase'

function Input(){
  const [task, setTask] = useState()   // <------------ creating state variable

  function handleChange(e){
    setTask(e.target.value)    // <---------- updating the state value
  }

  return(
    <div className="input">
      <input type="text" placeholder="Enter a Todo..." value={task} onChange={handleChange} />
      <button className="add-btn">
      <i className="fa fa-plus-circle" aria-hidden="true"></i><span id="btn-text">Add Todo</span> 
      </button>
    </div>
  )
}

export default Input
```

### Create operation

Let's add the entered data into the database when the user clicks the `Add Todo` button.

> Add a `onClick` attributewhich invokes `handleClick` function.

Reviewer: grammar and typo

```js
function handleClick() {
  const todoRef = firebase.database().ref('todo')  // <---------- creating a reference in database
  // code goes here
  }
```

To add data to the database, you can use `set()` to save data to a specified reference, replacing any existing data at that path. 

We use the `push()` method to append data to a list in multiuser applications. The `push()` method generates a unique key every time a new child is added to the specified Firebase reference. 

**More Info:** By using these auto-generated keys for each new element in the list, several clients can add children to the same location simultaneously without writing conflicts. The unique key generated by `push()` is based on a timestamp, so list items are automatically ordered chronologically.

⚠ **Note**: Using `set()` overwrites data at the specified location, including any child nodes.

```js
function handleClick() {
  const todoRef = firebase.database().ref('todo')  
  const newTodoRef = todoRef.push()
  const todo = {
      task,
      completed: false
    }
  newTodoRef.set(todo)
  setTask("")             
}
```

> Here 👆, we added the data that we get as input into the database. By default, the task was incomplete. After the data is added to the database, the input field is cleared by setting `task` to an empty string(`setTask("")`).

Reviewer: ```suggestion
Here 👆, we added the data that we get as input into the database. By default, the task is incomplete. After the data is added to the database, the input field is cleared by setting `task` to an empty string(`setTask("")`).
```

<details>

> <summary>Your final Input.js will look like this.</summary>

Reviewer: ```suggestion
<summary>Your final code in `Input.js` should look like this.</summary>
```

function Input(){
  const [task, setTask] = useState('')

  function handleChange(e){
    setTask(e.target.value)
  }
  function handleClick(){
    const todoRef = firebase.database().ref('todo')  
    const newTodoRef = todoRef.push()
    const todo = {
      task,
      completed: false
    }
    newTodoRef.set(todo)
    setTask("")
  }

  return(
    <div className="input">
      <input type="text" placeholder="Enter a Todo..." value={task} onChange={handleChange} />
      <button className="add-btn" onClick={handleClick}>
      <i className="fa fa-plus-circle" aria-hidden="true"></i><span id="btn-text">Add Todo</span> 
      </button>
    </div>
  )
}

export default Input
```

</details>

Now import `Input.js` into `App.js` and check your output by adding some data.

```jsx
import React from 'react'
import Input from './components/Input'   // <----------- importing Input.js
import './App.css'

function App() {
  return (
    <div className="App">
      <h1>Todo App</h1>
      <Input />                        // <------------ rendering the component
    </div>
  );
}

export default App;
```

![Output with input](https://cloud-792qf1oid.vercel.app/0input.js.png)

> Looks good, doesn't it! 😂

Reviewer: emoji is kinda inappropriate for this sentence?

![Database](https://cloud-792qf1oid.vercel.app/1input_to_database.png)

> The data you add through `input`, after clicking the button, will be appended to Firebase in real-time.

Reviewer: ```suggestion
The data you added through the `input` will be appended to our Firebase Database!
```

> ### Creating Todo List

Reviewer: NEXT REVIEW FROM HERE:

Create a new component, `TodoList.js`, in the components folder. Import `React`, `Firebase`, create and export a functional component. 

```jsx
import React, { useState, useEffect } from 'react'
import firebase from '../Firebase'

function TodoList(){
  // code goes here
  return(
    <div className="todo-list">

    </div>
}

export default TodoList
```

### Read operation

We will then read the data in the database, fetch it into an array, and display it on the screen.

Create a state variable `todoList`. This will be the empty array to which we add the data fetched from the database.

```jsx
 const [todoList, setTodoList] = useState([])           // <-------- state variable 
```

We will read the data only once after the component is rendered. To do that, we use the `useEffect` hook.

```jsx
useEffect(() => {
  // code goes here
}, [])
```

> We have got the data in a specified reference. 

Reviewer: ```suggestion
We have our the data stored in a specified reference. 
```

> To read data at a path and listen for changes, use the `on()` method of `firebase.database.Reference` to observe events. The `on()` method is used to synchronize data in real-time. 

Reviewer: ```suggestion
To read data at a path and listen for changes, we will use the `on()` method of `firebase.database().ref()` to observe events. The `on()` method is used to synchronize data in real-time. 
```

```jsx
useEffect(() => {
  const todoRef = firebase.database().ref('todo')
  // Syncronizing
  todoRef.on()

}, [])
```

The `on()` method takes two parameters-
1) The event type - how you control the level of synchronization from the real-time database.
1) The callback function - the event type above controls the callback function.

Here, we are going to use the `value` event.

| Event  | Typical usage  | 
| :------------- | :----------: |
| `value` | Read and listen for changes to the entire contents of a path.  |
<br/>

The callback function is called every time there is a change at the specified location in the database.  The event callback is passed a `snapshot` containing all data at that location, including child data. If there is no data, the snapshot will return `null` when you call `val()` on it.

```jsx
todoRef.on("value", (snapshot) => {
  const todos = snapshot.val()          // <---------- value of the data snapshot
})
```

> ⭐ **Important**: The `value` event is called every time data is changed at the specified database reference, including changes to children. To limit the size of your snapshots, attach only at the lowest level needed for watching changes. For example, attaching a listener to the root of your database is not recommended.

Reviewer: confusing, what do we "attach"?

Now we need the data as an array to put it into the state. Declare an empty array and add the `id` and `task` into the array.

```jsx
todoRef.on("value", (snapshot) => {
  const todos = snapshot.val()
  const todoList = []
  for(let id in todos){
    todoList.push({id,...todos[id]})            // <------------ adding todos to todoList
  }
  setTodoList(todoList)         // <----------- setting the state
})
```

Now that we have the data in `todoList`, we can map through it to display it.

Initially, it is an empty array. So we have to check whether it is empty or not, then map through it. 

> For now, let's display using `<h1>`. Later we create a component for saying it.

Reviewer: ```suggestion
For now, let's display our list using the `<h1>` tag but later, we will create a separate component for rendering it.
```

```jsx
return(
  <div className="todo-list">
    {todoList ? todoList.map((todo) => <h1> {todo.task}</h1>) : null}
  </div>
)
```

<details>

> <summary>Your TodoList component so far looks like:</summary>

Reviewer: ```suggestion
<summary>Your `TodoList` component's code so far:</summary>
```

    })

  }, [])
  
  return(
    <div className="todo-list">
      {todoList ? todoList.map((todo) => <h1> {todo.task}</h1>) : null}
    </div>
  )
}

export default TodoList
```

</details>   

This will display the data that is retrieved from the database. Make sure you import and render the `TodoList` component into `App.js`.

![Retrived data](https://cloud-792qf1oid.vercel.app/2reading_database.png)

So far, we have done **Create** and **Read** operations. Halfway done!!! 🙌

![Half way done](https://cloud-792qf1oid.vercel.app/7half_way_done_.gif)

Let's continue to work with **Update** and **Delete** operations.

### Creating Todo Item

Create another file, `Todo.js`, in the `components` folder. Import React and Firebase. Also, create a functional component that takes a prop `todo`.

```jsx
import React from 'react'.
import firebase from '../Firebase'

function Todo({ todo }){
  return(
    <div classsName="todo-item">

    </div>
  )
}

export default Todo
```

Update the `TodoList` by importing `Todo` and rendering it.

```jsx
return(
  <div className="todo-list">
    {todoList ? todoList.map((todo, i) => <Todo todo={ todo } key={ i } />) : null}
  </div>

> )

Reviewer: can you please explain what we are doing here. Because, i know that we are rendering a todo component for each todo and also passing the data to that component as prop, but beginners will find it confusing, please add a short explanation on this

      }
      
      setTodoList(todoList)
    })

  }, [])
  
  return(
    <div className="todo-list">
      {todoList ? todoList.map((todo, i) => <Todo todo={ todo } key={ i } />) : null}    // <----------- updated with 
      <Todo />
    </div>
  )
}

export default TodoList
```

</details>

Hop on to the `Todo` component, and let's display the data.

```jsx
function Todo({ todo }){
  return(
    <div className="todo-item">
      <div className="task">
        <input type="checkbox" />
        <p>{todo.task}</p>
      </div>
      <div className="buttons">
        <button className="del-btn"><i className="fa fa-trash" aria-hidden="true"></i></button>
      </div>
    </div>
  )
}
```

This displays the todo items with a `checkbox` and `delete` button.

Let's add functionality to the buttons.

### Update operation

When the `checkbox` is checked, the task is completed. Let's update the value of `completed` to `true`.

Add the `onClick` attribute to the checkbox that invokes `completeTodo` function. Whether the `checkbox` is checked or not depends on the `completed` value.

```jsx
<input type="checkbox" onClick={ completeTodo } checked={ todo.completed }/>
```

> To simultaneously write to specific children of a node without overwriting other child nodes, we can use the `update()` method. When calling `update()` updates lower-level child values by specifying a path for the key.

Reviewer: ```suggestion
To simultaneously write to specific children of a node without overwriting other child nodes, we can use the `update()` method. When we call `update()`, it updates lower-level child values by specifying a path for the key.
```

    todoRef.child(todo.id).update({
      completed: !todo.completed
    })
  }
  return(
    <div className="todo-item">
      <div className="task">
        <input type="checkbox" onClick={ completeTodo }checked={todo.completed}/>
        <p style={todo.completed ? completedStyle : null} >{todo.task}</p>               // <------------ adding className depending on todo.completed
      </div>
      <div className="buttons">
        <button className="del-btn"><i className="fa fa-trash" aria-hidden="true"></i></button>
      </div>
    </div>
  )
}
```

</details>

Now our completed tasks will be styled accordingly.

![struck off](https://cloud-792qf1oid.vercel.app/5struckoff.png)

Let's add functionality to our `delete` button.

### Delete operation

Add `onClick` attribute to the button with a `deleteTodo` function.

```jsx
<button className="del-btn" onClick={ deleteTodo }>
  <i className="fa fa-trash" aria-hidden="true"></i>
</button>
```

The simplest way to delete data is to call `remove()` on a reference to that data location.

```jsx
function deleteTodo(){
  const todoRef = firebase.database().ref('todo')
  todoRef.child(todo.id).remove()
}
```

<details>

> <summary>Our final Todo component will be somethink like this:</summary>

Reviewer: ```suggestion
<summary>Our final `Todo` component will be something like this:</summary>
```

        opacity: 0.4,
        textDecoration: "line-through"
    }
  
  function completeTodo(){
    const todoRef = firebase.database().ref('todo')
    todoRef.child(todo.id).update({
      completed: !todo.completed
    })
  }
  function deleteTodo(){
    const todoRef = firebase.database().ref('todo')
    todoRef.child(todo.id).remove()
  }

  return(
    <div className="todo-item">
      <div className="task">
        <input type="checkbox" onClick={ completeTodo }checked={todo.completed}/>
        <p style={todo.completed ? completedStyle:null} >{todo.task}</p>
      </div>
      <div className="buttons">
        <button className="del-btn" onClick={ deleteTodo }><i className="fa fa-trash" aria-hidden="true"></i></button>
      </div>
    </div>
  )
}

export default Todo
```

</details>

![Delete todo](https://cloud-792qf1oid.vercel.app/4deleting_database.png)

That's it, folks! We completed our Full-stack Todo App. Check your final code [here](https://repl.it/@Giridharhackclu/Todo-App-Final#src/App.js). ✌

![Completed](https://cloud-792qf1oid.vercel.app/6you_did_it_.gif)

## Hacking

Now it's your turn to customize.

* Go wild with the styles. Create your custom Todo Apps.

> * Add more functionality.

Reviewer: please specify some examples

* Filter the tasks as completed and due.

> * Try to add a database to your projects.

Reviewer: aren't we already using a database?

> I recommend customizing the project in this workshop. Share your creation with everyone in [#ship](https://hackclub.slack.com/messages/ship) in hack club slack. Do share with me too!!! I'm [@Giridhar](https://hackclub.slack.com/team/U013E6KE9UJ) on slack. I'd love to hear from you. 

Reviewer: ```suggestion
I recommend customizing this project as much as you can! Share your creation with everyone in the [#ship](https://hackclub.slack.com/messages/ship) channel on the Hack Club slack. Don't forget to share it with me too!!! I'm [@Giridhar](https://hackclub.slack.com/team/U013E6KE9UJ) on slack. I'd love to hear from you. 
```
